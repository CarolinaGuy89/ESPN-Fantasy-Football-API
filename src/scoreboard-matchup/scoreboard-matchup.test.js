import _ from 'lodash';

import BaseObject from '../base-object/base-object.js';
import Team from '../team/team.js';
import ScoreboardMatchup from './scoreboard-matchup.js';

import { localObject, serverResponse } from './scoreboard-matchup.stubs.js';

describe('ScoreboardMatchup', () => {
  let leagueId;
  let seasonId;

  beforeEach(() => {
    leagueId = 513422;
    seasonId = 2017;
  });

  afterEach(() => {
    leagueId = null;
    seasonId = null;
  });

  test('extends BaseObject', () => {
    const instance = new ScoreboardMatchup();
    expect(instance).toBeInstanceOf(BaseObject);
  });

  describe('when creating a team from a server response', () => {
    test('parses and assigns data correctly', () => {
      const instance = ScoreboardMatchup.buildFromServer(serverResponse, { leagueId, seasonId });
      expect(instance).toMatchSnapshot();
    });
  });

  describe('when creating a team locally', () => {
    test('parses and assigns data correctly', () => {
      const instance = new ScoreboardMatchup(localObject);
      expect(instance).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    describe('when options are not passed', () => {
      const testPropIsUndefined = (prop) => {
        test(`${prop} is undefined`, () => {
          const newInstance = new ScoreboardMatchup();
          expect(_.get(newInstance, prop)).toBeUndefined();
        });
      };

      testPropIsUndefined('leagueId');
      testPropIsUndefined('seasonId');
    });

    describe('when options are passed', () => {
      const testPropIsSetFromOptions = (prop) => {
        test(`${prop} is set from options`, () => {
          const value = 25;
          const newInstance = new ScoreboardMatchup({ [prop]: value });
          expect(newInstance[prop]).toBe(value);
        });
      };

      testPropIsSetFromOptions('leagueId');
      testPropIsSetFromOptions('seasonId');
    });
  });

  describe('responseMap', () => {
    const buildScoreboardMatchup = (data, options) => (
      ScoreboardMatchup.buildFromServer(data, options)
    );

    const testTeamBehavior = ({ isTeamHome }) => {
      const teamPrefix = isTeamHome ? 'home' : 'away';

      describe('when the team data does not exist on the response', () => {
        test(`sets ${teamPrefix}Team to undefined`, () => {
          const data = { teams: [] };

          const returnedMatchup = buildScoreboardMatchup(data);
          expect(_.get(returnedMatchup, `${teamPrefix}Team`)).toBeUndefined();
        });
      });

      describe('when the team data does exist on the response', () => {
        describe('when a matching team is in the cache', () => {
          test(`sets ${teamPrefix}Team to the matching cached team`, () => {
            const teamId = 5;
            const data = {
              teams: [{
                teamId,
                team: { teamId },
                home: isTeamHome
              }, {
                teamId: teamId + 1,
                team: { teamId: teamId + 1 },
                home: !isTeamHome
              }]
            };

            const cachedTeam = Team.buildFromServer({ teamId }, { leagueId, seasonId });

            const returnedMatchup = buildScoreboardMatchup(data, { leagueId, seasonId });
            expect(_.get(returnedMatchup, `${teamPrefix}Team`)).toBe(cachedTeam);
          });
        });

        describe('when a matching team is not in the cache', () => {
          test(`sets ${teamPrefix}Team to a new team instance`, () => {
            const teamId = 5;
            const data = {
              teams: [{
                teamId,
                team: { teamId },
                home: isTeamHome
              }, {
                teamId: teamId + 1,
                team: { teamId: teamId + 1 },
                home: !isTeamHome
              }]
            };

            const returnedMatchup = buildScoreboardMatchup(data, { leagueId, seasonId });
            expect(_.get(returnedMatchup, `${teamPrefix}Team`)).toEqual(
              Team.buildFromServer(data.teams[0], { leagueId, seasonId })
            );
          });
        });
      });
    };

    describe('homeTeam', () => {
      describe('manualParse', () => {
        testTeamBehavior({ isTeamHome: true });
      });
    });

    describe('awayTeam', () => {
      describe('manualParse', () => {
        testTeamBehavior({ isTeamHome: false });
      });
    });

    const testTeamScoreBehavior = ({ isTeamHome }) => {
      const teamPrefix = isTeamHome ? 'home' : 'away';

      describe('when there is matching team in the response data', () => {
        test(`sets ${teamPrefix}TeamScore to the correct score`, () => {
          const score = 213.43;
          const data = {
            teams: [{
              score,
              home: isTeamHome
            }, {
              score: score * 0.75,
              home: !isTeamHome
            }]
          };

          const returnedMatchup = buildScoreboardMatchup(data, { leagueId, seasonId });
          expect(_.get(returnedMatchup, `${teamPrefix}TeamScore`)).toBe(score);
        });
      });

      describe('when there is not matching team in the response data', () => {
        test(`sets ${teamPrefix}TeamScore to undefined`, () => {
          const data = {
            teams: [{
              score: 213.43,
              home: !isTeamHome
            }]
          };

          const returnedMatchup = buildScoreboardMatchup(data, { leagueId, seasonId });
          expect(_.get(returnedMatchup, `${teamPrefix}TeamScore`)).toBeUndefined();
        });
      });
    };

    describe('homeTeamScore', () => {
      describe('manualParse', () => {
        testTeamScoreBehavior({ isTeamHome: true });
      });
    });

    describe('awayTeamScore', () => {
      describe('manualParse', () => {
        testTeamScoreBehavior({ isTeamHome: false });
      });
    });

    describe('winner', () => {
      const callWinner = (response, instance) => ScoreboardMatchup.responseMap.winner.manualParse(
        undefined, response, undefined, instance
      );

      describe('when the winner is the home team', () => {
        test('returns the parsed homeTeam', () => {
          const response = { winner: 'home' };
          const homeTeam = new Team();
          const instance = new ScoreboardMatchup({ homeTeam });

          const returnedWinner = callWinner(response, instance);
          expect(returnedWinner).toBe(homeTeam);
        });
      });

      describe('when the winner is the away team', () => {
        test('returns the parsed awayTeam', () => {
          const response = { winner: 'away' };
          const awayTeam = new Team();
          const instance = new ScoreboardMatchup({ awayTeam });

          const returnedWinner = callWinner(response, instance);
          expect(returnedWinner).toBe(awayTeam);
        });
      });

      describe('when there is no winner', () => {
        test('returns undefined', () => {
          const response = { winner: '' };
          const homeTeam = new Team();
          const awayTeam = new Team();
          const instance = new ScoreboardMatchup({ homeTeam, awayTeam });

          const returnedWinner = callWinner(response, instance);
          expect(returnedWinner).toBeUndefined();
        });
      });
    });
  });
});
