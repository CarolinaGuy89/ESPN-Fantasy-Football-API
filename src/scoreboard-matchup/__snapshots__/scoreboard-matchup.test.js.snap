// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreboardMatchup when creating a team from a server response parses and assigns data correctly 1`] = `
ScoreboardMatchup {
  "awayTeam": Team {
    "abbreviation": "GOAT",
    "awayLosses": 1,
    "awayTies": 0,
    "awayWins": 5,
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionLosses": 2,
    "divisionTies": 0,
    "divisionWinningPercentage": 66.667,
    "divisionWins": 4,
    "firstName": "The Boston",
    "homeLosses": 3,
    "homeTies": 0,
    "homeWins": 4,
    "lastName": "TE Party",
    "leagueId": undefined,
    "logoURL": "http://assets.nydailynews.com/polopoly_fs/1.1443336.1378094427!/img/httpImage/image.jpg_gen/derivatives/article_750/sport-nfl-football.jpg",
    "losses": 4,
    "pointsAgainst": 1131.4,
    "pointsFor": 1305.78,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "W",
    "teamId": 1,
    "ties": 0,
    "waiverRank": 6,
    "winningPercentage": 69.231,
    "wins": 9,
  },
  "awayTeamScore": 103.84,
  "homeTeam": Team {
    "abbreviation": "GRON",
    "awayLosses": 2,
    "awayTies": 0,
    "awayWins": 4,
    "division": Object {
      "divisionId": 1,
      "divisionName": "Division 2",
      "size": 4,
    },
    "divisionLosses": 4,
    "divisionTies": 0,
    "divisionWinningPercentage": 33.333,
    "divisionWins": 2,
    "firstName": "Prestige",
    "homeLosses": 6,
    "homeTies": 0,
    "homeWins": 1,
    "lastName": "Worldwide",
    "leagueId": undefined,
    "logoURL": "http://3219a2.medialib.glogster.com/media/f8/f81cfe4e18dba8da78f1e654f311d7852eedf937770ef0670f656835288409ca/prestige-png.png",
    "losses": 8,
    "pointsAgainst": 1232.94,
    "pointsFor": 1062.82,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "L",
    "teamId": 2,
    "ties": 0,
    "waiverRank": 5,
    "winningPercentage": 38.462,
    "wins": 5,
  },
  "homeTeamScore": 56.64,
  "isByeWeek": false,
  "leagueId": undefined,
  "seasonId": undefined,
  "winner": Team {
    "abbreviation": "GOAT",
    "awayLosses": 1,
    "awayTies": 0,
    "awayWins": 5,
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionLosses": 2,
    "divisionTies": 0,
    "divisionWinningPercentage": 66.667,
    "divisionWins": 4,
    "firstName": "The Boston",
    "homeLosses": 3,
    "homeTies": 0,
    "homeWins": 4,
    "lastName": "TE Party",
    "leagueId": undefined,
    "logoURL": "http://assets.nydailynews.com/polopoly_fs/1.1443336.1378094427!/img/httpImage/image.jpg_gen/derivatives/article_750/sport-nfl-football.jpg",
    "losses": 4,
    "pointsAgainst": 1131.4,
    "pointsFor": 1305.78,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "W",
    "teamId": 1,
    "ties": 0,
    "waiverRank": 6,
    "winningPercentage": 69.231,
    "wins": 9,
  },
}
`;

exports[`ScoreboardMatchup when creating a team locally parses and assigns data correctly 1`] = `
ScoreboardMatchup {
  "awayTeam": Team {
    "abbreviation": "GOAT",
    "awayLosses": 1,
    "awayTies": 0,
    "awayWins": 5,
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionLosses": 2,
    "divisionTies": 0,
    "divisionWinningPercentage": 0.66667,
    "divisionWins": 4,
    "firstName": "The Boston",
    "homeLosses": 3,
    "homeTies": 0,
    "homeWins": 4,
    "lastName": "TE Party",
    "leagueId": undefined,
    "logoURL": "http://assets.nydailynews.com/polopoly_fs/1.1443336.1378094427!/img/httpImage/image.jpg_gen/derivatives/article_750/sport-nfl-football.jpg",
    "losses": 4,
    "pointsAgainst": 1131.4,
    "pointsFor": 1305.78,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "W",
    "teamId": 1,
    "ties": 0,
    "waiverRank": 6,
    "winningPercentage": 0.69231,
    "wins": 9,
  },
  "awayTeamScore": 103.84,
  "homeTeam": Team {
    "abbreviation": "GRON",
    "awayLosses": 2,
    "awayTies": 0,
    "awayWins": 4,
    "division": Object {
      "divisionId": 1,
      "divisionName": "Division 2",
      "size": 4,
    },
    "divisionLosses": 4,
    "divisionTies": 0,
    "divisionWinningPercentage": 0.33333,
    "divisionWins": 2,
    "firstName": "Prestige",
    "homeLosses": 6,
    "homeTies": 0,
    "homeWins": 1,
    "lastName": "Worldwide",
    "leagueId": undefined,
    "logoURL": "http://3219a2.medialib.glogster.com/media/f8/f81cfe4e18dba8da78f1e654f311d7852eedf937770ef0670f656835288409ca/prestige-png.png",
    "losses": 8,
    "pointsAgainst": 1232.94,
    "pointsFor": 1062.82,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "L",
    "teamId": 2,
    "ties": 0,
    "waiverRank": 5,
    "winningPercentage": 0.38462,
    "wins": 5,
  },
  "homeTeamScore": 56.64,
  "isByeWeek": false,
  "leagueId": undefined,
  "seasonId": undefined,
  "winner": Team {
    "abbreviation": "GOAT",
    "awayLosses": 1,
    "awayTies": 0,
    "awayWins": 5,
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionLosses": 2,
    "divisionTies": 0,
    "divisionWinningPercentage": 0.66667,
    "divisionWins": 4,
    "firstName": "The Boston",
    "homeLosses": 3,
    "homeTies": 0,
    "homeWins": 4,
    "lastName": "TE Party",
    "leagueId": undefined,
    "logoURL": "http://assets.nydailynews.com/polopoly_fs/1.1443336.1378094427!/img/httpImage/image.jpg_gen/derivatives/article_750/sport-nfl-football.jpg",
    "losses": 4,
    "pointsAgainst": 1131.4,
    "pointsFor": 1305.78,
    "seasonId": undefined,
    "streakLength": 1,
    "streakType": "W",
    "teamId": 1,
    "ties": 0,
    "waiverRank": 6,
    "winningPercentage": 0.69231,
    "wins": 9,
  },
}
`;
